### ISSUE TITLE
Support for free-threading CPython

### ISSUE THREAD

**Issue by neutrinoceros** [h5py/h5py#2475]
As hinted to in #2416, I think there are a couple small things we can do to help the community of downstream testers to run h5py on free-threaded CPython. I just have a couple minutes on my hands so instead of *doing* them, I'll just write a list for now.
- review unsafe C-API usage (most importantly use of borrowed refs), see https://docs.python.org/3.13/howto/free-threading-extensions.html
- declare support in all C and Cython extensions (see https://py-free-threading.github.io/porting/ and https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiler-directives)
- setup some minimal validation in CI (like running tests in a single thread on Python 3.13t on ubuntu and check that everything passes and the GIL stays disabled)

I am hoping this is a relatively lightweight job, so right now I think that doing all this in time for h5py 3.12 is feasible, so let me put it in the milestone.

**Comment by ajelenak** [comment]
What implications of running h5py with free-threaded CPython will be on libhdf5 which is not thread-safe by default? I don't think h5py currently uses thread-safe build of libhdf5.

**Comment by tacaswell** [comment]
We already have our own global lock on any access to libhdf5 https://github.com/h5py/h5py/blob/7d4c3a15fc53f7d0a5b6c8e9895a9b3fb9632f12/h5py/_objects.pyx#L20-L46.  The only place I am currently worried is in the code around the non-local close registry https://github.com/h5py/h5py/blob/7d4c3a15fc53f7d0a5b6c8e9895a9b3fb9632f12/h5py/_objects.pyx#L95-L98 but it looks like everything is already behind `phil`

**Comment by neutrinoceros** [comment]
Given the discussion in #2478 I think I'd like to wait on Cython 3.1 before making any more progress on this. In the mean time, it shouldn't be blocking the 3.12 release.

**Comment by tacaswell** [comment]
I _think_ it is also reasonable to build cp313t wheels, but leave the "not freethread safe" state on it.  That way downstream can still get pre-built wheels and then turn the safety checks off to test.

**Comment by neutrinoceros** [comment]
How about we do this but just for nightlies for now ?

**Comment by neutrinoceros** [comment]
So that's basically what I tried in #2449, but the reason I abandoned it is that I think Cython might soon have a stable release that supports cp313t. If Cython 3.1 is not out in a month we can revive #2449, sounds good ?

**Comment by SerGeRybakov** [comment]
Hi everyone! Can you, please, explain the necessity of `phil` locks? As I understand, it behaves like python GIL, but I don't get what for and why. I tried to switch it off, but even during reading I get Segmentation Faults in files, when each file is being read by a single thread from a `ThreadPoolExecutor`.

**Comment by tacaswell** [comment]
libhd5 is not thread safe around the internal state it maintains about the files it has open (things like caches, reference counts, etc).  

The point of the `phil` is to prevent the segfaults you note ;)

**Comment by tacaswell** [comment]
> I think it is also reasonable to build cp313t wheels, but leave the "not freethread safe" state on it

Numpy is going the otherway and optimistically setting the flag and assuming anyone using the 313t builds is knowingly signing up for some segfault discovery.

**Comment by takluyver** [comment]
And to be clear, even when h5py supports free-threading Python, there will still be a lock protecting HDF5 operations. So you won't get two threads reading/writing HDF5, but you could do HDF5 things in one thread and something else in a separate thread (in fact you already can to some extent, because h5py releases the GIL, but it's more limited).

This is a limitation that comes from HDF5 itself, there's no way that I know for h5py to bypass it.

**Comment by 0xDEADFED5** [comment]
looks like ya'll are pretty close to it, eh?

here's another vote for 3.13t wheels =)

**Comment by neutrinoceros** [comment]
The current blocker is that we want to setup a CI job running our test suite with `PYTHON_GIL=0`, which should become much easier soon, when `actions/setup-python` adds support for free-threaded build in their next release.

**Comment by grantlouisherman** [comment]
Hey has there been any movement on this? I wanted to use HD5Py in free threaded and Im interested in this progress

**Comment by neutrinoceros** [comment]
We were blocked by Cython until recently, now #2478 is waiting for reviews.

**Comment by crusaderky** [comment]
I'm actively working on this.

Open PRs:
- #2650 
- #2655 

Downstream of the two PRs above, reaching a state where `pytest h5py --parallel-threads=32` is consistently green would demonstrate with a good degree of confidence that multiple threads can operate on the same h5py.File, as long as you don't apply mutating commands to the same dataset/group/attribute in parallel.
Note that this is a much stronger thread safety guarantee than proving that separate threads can operate on separate files.

The current situation, however, is that with just 2 parallel threads h5py deadlocks hard and fast.

### Oustanding work:

- [x] fix actual race conditions in h5py, until the tests from #2655 are robustly green in pytest-run-parallel
- [x] produce 3.14t wheels (#2702)
- [x] #2655
- [x] add pytest-run-parallel to CI https://github.com/h5py/h5py/pull/2761
- [ ] add specific tests for mutating the same groups/datasets/attributes from different threads, and fix more race conditions that may arise
- [x] run test suite under TSAN. Reach a green TSAN run. 
- [ ] (optional) add TSAN to CI.
- [x] set the freethreading flag in Cython (#2745)
- [ ] bump up pypi marker to 2 - Beta, update documentation
- [ ] rest.

**Comment by crusaderky** [comment]
# Status update
- As of master git tip as of today, h5py sets the freethreading-compatible flag (#2745)
- Next release will include 3.14t wheels (but not 3.13t) (#2702)
- ~There are no known race conditions~
- I have high confidence that there are no unknown race conditions when different threads operate on separate `h5py.Dataset`'s within the same `h5py.File` (#2655)
- I have LOW confidence that there are no unknown race conditions when different threads operate on the same `h5py.Dataset` in parallel (untested)

**Comment by crusaderky** [comment]
Update: https://github.com/h5py/h5py/pull/2761 highlights two race conditions that are invisible in pytest-run-parallel until one moves from localhost to CI. I'm working to reproduce them (and possibly more) with TSAN/ASAN.

**Comment by crusaderky** [comment]
Update: all race conditions when two threads share the same File (but not the same Dataset) have been fixed in main. https://github.com/h5py/h5py/pull/2761 is awaiting review.

Currently busy writing extra tests for when two threads share the same Dataset.

**Comment by crusaderky** [comment]
# Status update
pytest-run-parallel is now in the main branch!
As of today I have
- *very high* confidence that there are no race conditions when two threads operate on the same File, but on different Datasets
- *low* confidence that there are no race conditions when two threads operate on the same Dataset

## Outstanding work

- [ ] Add specific tests for mutating the same groups/datasets/attributes from different threads, and fix more race conditions that may arise
- [ ] (needs discussion) add TSAN to CI. Today this is already done by https://github.com/crusaderky/hdf5-pixi, and it implies quite a bit of faff. I am not very convinced that running it in the h5py main CI has enough value to justify the effort.
- [ ] update documentation
- [ ] (needs discussion) bump up pypi marker to 2 - Beta or 3 - Stable
